### PR TITLE
Answer:6

### PR DESCRIPTION
**Pull request by BruceArmstrong007**
## Checklist for challenge submission

- [x] Start your PR message with Answer:${challenge_number}

Added and implemented canMatch guard & hasRole directive

### CODE REVIEW

File: apps/angular/permissions/src/app/common/guards/is-admin.guard.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import { CanMatchFn } from '@angular/router';
+import { UserStore } from '../../user.store';
+import { map } from 'rxjs';
+import { TypedRoute } from '../../routes';
+
+export const HasPermission: CanMatchFn = (route: TypedRoute, segments) => {
+  const user$ = inject(UserStore).user$;
+  return user$.pipe(
+    map((user) => {
+      if (route?.data?.isAdmin) return user?.isAdmin ? true : false;
```
Comment (tomalaforge) [inline apps/angular/permissions/src/app/common/guards/is-admin.guard.ts:11]: don't need the true and false.

File: apps/angular/permissions/src/app/common/guards/is-admin.guard.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      else {
```
Comment (tomalaforge) [inline apps/angular/permissions/src/app/common/guards/is-admin.guard.ts:12]: don't need the else, you already have a early return

File: apps/angular/permissions/src/app/routes.ts [hunk 3]
```diff
@@ -1,14 +1,78 @@
-export const APP_ROUTES = [
+import { Route } from '@angular/router';
+import { Role } from './user.model';
+import { HasPermission } from './common/guards/is-admin.guard';
+export interface TypedRoute extends Route {
+  data?: {
+    isAdmin?: boolean;
+    roles?: Role[];
+  };
+}
```
Comment (tomalaforge) [inline apps/angular/permissions/src/app/routes.ts:9]: 👍

File: apps/angular/permissions/src/app/routes.ts [hunk 4]
```diff
@@ -1,14 +1,78 @@
...
+export const APP_ROUTES: TypedRoute[] = [
   {
     path: '',
     loadComponent: () =>
       import('./login.component').then((m) => m.LoginComponent),
   },
   {
     path: 'enter',
+    canMatch: [HasPermission],
```
Comment (tomalaforge) [inline apps/angular/permissions/src/app/routes.ts:18]: HasPermission is a function which can take parameters. Don't use the data property

File: apps/angular/permissions/src/app/common/guards/is-admin.guard.ts (hunk 1 above)
Comment (BruceArmstrong007) [reply apps/angular/permissions/src/app/common/guards/is-admin.guard.ts:11]: yes, it doesn't need to be , i will fix it

File: apps/angular/permissions/src/app/routes.ts (hunk 4 above)
Comment (BruceArmstrong007) [reply apps/angular/permissions/src/app/routes.ts:18]: @tomalaforge  u mean i can call that function directly as well ?

Comment (tomalaforge) [reply apps/angular/permissions/src/app/routes.ts:18]: That's the goal of the new API. Guard are function, not class,

Comment (BruceArmstrong007) [reply apps/angular/permissions/src/app/routes.ts:18]: i will update that.

**Pull request closed without merging**